### PR TITLE
[Woo POS] collect payment on new card present payment service

### LINKED ISSUE #12867: [Woo POS] Create Payment Adaptor
Initially, this should include:
- Creating the `CardPresentPaymentPreflightController`
- Creating the `CollectOrderPaymentUseCase`
- Propagating alerts to the common publisher by implementing `CardPresentPaymentAlertsPresenting`
- Calling `collectPayment` for the order passed in

### PR DESCRIPTION
**Pull request by joshheald**
Closes: #12867
Merge after: #12944 

## Description

In this PR we add the Collect Payment adaptor, which uses the `CollectOrderPaymentUseCase` to take a payment.

I've also linked the `Card` button to use this to take payment for a test order of $15, so we can see it working in practice.

This is primarily happy path – we'll need to focus on error handling properly later, and #12496 continues to implement the button handlers properly, but will follow this.

## Steps to reproduce

1. Launch the app using a US-based WooPayments store eligilble for POS
2. Enable the POS mode if it's not already
3. Navigate to `Menu > Point of Sale Mode`
4. Select a product
5. Tap `Checkout`
6. Tap `Card`
7. Observe that the payment process starts, connecting a reader if required, and completes with a paid order
8. Check the `Orders` tab to see your order and that it's been paid for by card.

## Testing information

I've tested this with a WPCom store and a self-hosted (Pressable) store.

Cancellation isn't handled fully yet, so if you cancel a payment and then start another one, it might not work.

As with reader connection, we'll need to ensure robust handling when connect payment is attempted while a reader connection is ongoing. Again... happy path!

In particular, note that the buttons on the success modal don't work right at the moment. Swipe down to dismiss it. They'll be fixed in #12946 

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/2472348/213fb2c9-2c2a-43d2-b138-a3332673a988

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift [hunk 1]
```diff
@@ -74,12 +76,41 @@ final class CardPresentPaymentService: CardPresentPaymentFacade {
...
+            // TODO: fetch the receipt URL to return an accurate value.
+            let transaction = CardPresentPaymentTransaction(receiptURL: URL(string: "https://example.com")!)
+            return .success(transaction)
+        case .cancellation:
+            return .cancellation
+        }
     }
 
     func cancelPayment() {
+        self.paymentTask?.cancel()
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift]: super nit:
```suggestion
        paymentTask?.cancel()
```

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import Foundation
+import WooFoundation
+import Combine
+import struct Yosemite.Order
+import struct Yosemite.CardPresentPaymentsConfiguration
+
+struct CardPresentPaymentCollectOrderPaymentUseCaseAdaptor {
+    private let currencyFormatter = CurrencyFormatter(currencySettings: ServiceLocator.currencySettings)
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: nit: would be nice to DI this to enable unit testing

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift [hunk 3]
```diff
@@ -74,12 +76,41 @@ final class CardPresentPaymentService: CardPresentPaymentFacade {
...
-        // TODO: replace it with payment collection
-        .cancellation
+        paymentTask?.cancel()
+
+        // What happens if `start` gets called while there's a connection ongoing but not finished?
+        // Ideally, we'd adopt the connection attempt.
+        // Since we're reusing the connection controllers, that's a good start, but it needs proper testing.
+        let preflightController = createPreflightController()
+
+        // TODO: Update the connected reader subject when we get a connection here.
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift:88]: When we work on the [card reader disconnection support](https://github.com/orgs/woocommerce/projects/247/views/6?sliceBy%5Bvalue%5D=iOS+%F0%9F%8D%8E&pane=issue&itemId=65994779), we can consider observing the connected reader using `CardPresentPaymentAction.publishCardReaderConnections` to replace manual emitting events in `connectedReaderSubject` whenever a connection takes place.

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            }
+
+            let orderPaymentUseCase = CollectOrderPaymentUseCase(siteID: siteID,
+                                                 order: order,
+                                                 formattedAmount: formattedAmount,
+                                                 rootViewController: NullViewControllerPresenting(),
+                                                 onboardingPresenter: onboardingPresenter,
+                                                 configuration: configuration,
+                                                 alertsPresenter: alertsPresenter,
+                                                 preflightController: preflightController)
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: super nit: indentation

File: WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift [hunk 5]
```diff
@@ -77,6 +79,18 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
...
+
+    func cardPaymentTapped() {
+        Task { @MainActor in
+            showsCreatingOrderSheet = true
+            let order = try await createTestOrder()
+            showsCreatingOrderSheet = false
+            let _ = try await cardPresentPaymentService.collectPayment(for: order, using: .bluetooth)
+
+            // TODO: Here we should present something to show the payment was successful or not,
+            // and then clear the screen ready for the next transaction.
```
Comment (jaclync) [inline WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:91]: For reference, this is the design to show after a successful payment.

TfaZ4LUkEwEGrxfnEFzvJj/WooCommerce-POS?node-id=653-32958&t=MvK9jHYAj1UXyCVM-0 (design link shorthand didn't work for some reason)

Screenshot 2024-06-05 at 1 34 54 PM

File: WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift [hunk 6]
```diff
@@ -77,6 +79,18 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
     var checkoutButtonDisabled: Bool {
         return itemsInCart.isEmpty
     }
+
+    func cardPaymentTapped() {
```
Comment (jaclync) [inline WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:83]: Curious to get your thoughts to make it work for the latest design (likely the final for M1):

Screenshot 2024-06-05 at 1 50 57 PM

In this design, there's no CTA to take card payment if the reader is already connected (I'd think there is one CTA if the reader isn't connected). In this case (reader connected before landing on this checkout screen), do you think we'd just call `cardPresentPaymentService.collectPayment` right after knowing that the reader is connected on the checkout view? It's a good thing that we can decide how to show the UI now.

**Review by jaclync — APPROVED** [review]
Thanks for making everything async/await-friendly whenever possible!

Tested on a device with the reader and the flow worked nicely! 🎉 


https://github.com/woocommerce/woocommerce-ios/assets/1945542/fab05295-76cc-41ac-91af-ba7349368079

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift (hunk 1 above)
Comment (jaclync) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift]: Updated in https://github.com/woocommerce/woocommerce-ios/pull/12945/commits/4c9c1e4237387ceb6782b9baa2363aec9eab13a6

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: DI'ed in https://github.com/woocommerce/woocommerce-ios/pull/12945/commits/4c9c1e4237387ceb6782b9baa2363aec9eab13a6

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift (hunk 4 above)
Comment (jaclync) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: Updated in https://github.com/woocommerce/woocommerce-ios/pull/12945/commits/4c9c1e4237387ceb6782b9baa2363aec9eab13a6

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift (hunk 3 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift:88]: Yep, I agree... I think that will cover this requirement.

File: WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift (hunk 6 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:83]: Yes, exactly.

If the reader's connected, we start the payment when the screen loads and show a relevant message where "Tap or insert card to pay" is shown – I'd say we keep it simple and generally just say "preparing reader" until it's ready.

If the reader's not connected, I think we should show that flow modally to start with – but we could choose to only pop up the "connect to this reader" modal which they have to say yes or no to. Or show the buttons in-line.

Errors – tougher, as the merchant might be planning to take cash anyway, in which case it's annoying to have a card payment error pop up... but if they're taking a card payment, it needs to be obvious if that's failed for some reason, so showing a modal might be better there.

**Pull request merged**